project-koku/koku
# [COST-3344] OCP Volume Report does not respect group bys for capacity

## Jira Ticket

[COST-3344](https://issues.redhat.com/browse/COST-3344)

## Description

This change will adjust how capacity is calculated for OCP volume reports and will handle group by's appropriately. 

## Testing

1. Checkout Branch
2. Restart Koku
3. Create some test customer data with the `make create-test-customer` and `make load-test-customer-data` commands
4. Check the OCP volumes endpoint with a few different group bys and start/end dates and monthly resolutions and ensure the data matches up with what you see in the db:
- http://localhost:8000/api/cost-management/v1/reports/openshift/volumes/?group_by[cluster_id]=*&group_by[node]=*
- http://localhost:8000/api/cost-management/v1/reports/openshift/volumes/?group_by[node]=*&group_by[project]=* 
- http://localhost:8000/api/cost-management/v1/reports/openshift/volumes/?group_by[project]=*
- http://localhost:8000/api/cost-management/v1/reports/openshift/volumes/?group_by[cluster]=*&filter[resolution]=monthly&filter[time_scope_value]=-1&filter[time_scope_units]=month
- http://localhost:8000/api/cost-management/v1/reports/openshift/volumes/?group_by[node]=*&filter[resolution]=monthly&start_date=2022-12-15&end_date=2023-01-24
- http://localhost:8000/api/cost-management/v1/reports/openshift/volumes/?group_by[tag:app]=*
5. Some potentially helpful DB queries:
- this months data grouped by node, cluster_id and usage start: 
```
select sum(persistentvolumeclaim_capacity_gigabyte_months), usage_start, node, cluster_id from reporting_ocpusagelineitem_daily_summary where usage_start >= '2023-01-01' group by usage_start, node, cluster_id order by usage_start, cluster_id, node;
```
- this months data grouped by node and project:
```
select sum(persistentvolumeclaim_capacity_gigabyte_months), usage_start, node, namespace from reporting_ocpusagelineitem_daily_summary where usage_start >= '2023-01-01' group by usage_start, node, namespace order by usage_start, node, namespace;
```
- the sum of this months capacity grouped by cluster:
```
select sum(persistentvolumeclaim_capacity_gigabyte_months), cluster_id from reporting_ocpusagelineitem_daily_summary where usage_start >= '2023-01-01' group by cluster_id order by cluster_id;
```
- the sum of volume capacities for a specific day grouped by a specific tag keys values, in this case app:
```
select sum(persistentvolumeclaim_capacity_gigabyte_months), json_agg(volume_labels) from reporting_ocpusagelineitem_daily_summary where usage_start = '2023-01-16' and volume_labels ? 'app' group by volume_labels #>> '{app}'
```
## Notes

...
